src/devices/cpu/upd177x/upd177x.cpp

```diff
@@ -0,0 +1,797 @@
+	: upd177x_cpu_device(mconfig, UPD1771C, tag, owner, clock)
+{ }
+
+
+device_memory_interface::space_config_vector upd177x_cpu_device::memory_space_config() const
+{
+	return space_config_vector
+	{
+		std::make_pair(AS_PROGRAM, &m_program_config)
+	};
+}
+
+
+void upd177x_cpu_device::program_map(address_map &map)
+{
+	map(0, 0x1ff).rom();
+}
+
+
+void upd177x_cpu_device::device_start()
+{
+	space(AS_PROGRAM).specific(m_program);
+
+	state_add(UPD177X_PC,        "PC",    m_pc);
+	state_add(STATE_GENPC,       "GENPC", m_pc).noshow();
+	state_add(STATE_GENPCBASE,   "CURPC", m_pc).noshow();
+	state_add(UPD177X_SP,        "SP",    m_sp);
+	state_add(UPD177X_A,         "A",     m_a);
+	state_add(UPD177X_A_SAVE,    "A'",    m_a_save);
+	state_add(UPD177X_H,         "H",     m_h);
+	state_add(UPD177X_N,         "N",     m_n);
+	state_add(UPD177X_TIMER,     "Timer", m_timer);
+	state_add(UPD177X_X,         "X",     m_x);
+	state_add(UPD177X_Y,         "Y",     m_y);
+	state_add(UPD177X_PA,        "PA",    m_pa);
+	state_add(UPD177X_PB,        "PB",    m_pb);
+	state_add(UPD177X_MD0,       "MD0",   m_md0);
+	state_add(UPD177X_MD1,       "MD1",   m_md1);
+	state_add(UPD177X_PNC1,      "PNC1",  m_pnc1);
+	state_add(UPD177X_PNC2,      "PNC2",  m_pnc2);
+	state_add(UPD177X_SKIP,      "SKIP",  m_skip);
+	state_add(UPD177X_SKIP_SAVE, "SKIP'", m_skip_save);
+
+	save_item(NAME(m_pc));
+	save_item(NAME(m_sp));
+	save_item(NAME(m_a));
+	save_item(NAME(m_a_save));
+	save_item(NAME(m_h));
+	save_item(NAME(m_n));
+	save_item(NAME(m_timer));
+	save_item(NAME(m_x));
+	save_item(NAME(m_y));
+	save_item(NAME(m_pa));
+	save_item(NAME(m_pb));
+	save_item(NAME(m_ram));
+	save_item(NAME(m_md0));
+	save_item(NAME(m_md1));
+	save_item(NAME(m_pnc1));
+	save_item(NAME(m_pnc2));
+	save_item(NAME(m_skip));
+	save_item(NAME(m_skip_save));
+	save_item(NAME(m_ts));
+	save_item(NAME(m_ns));
+	save_item(NAME(m_ss));
+	save_item(NAME(m_dac));
+	save_item(NAME(m_dac_sign));
+	save_item(NAME(m_tn_int));
+	save_item(NAME(m_counter));
+
+	set_icountptr(m_icount);
+
+	m_channel = stream_alloc(0, 1, clock() / 4);
+}
+
+
+void upd177x_cpu_device::device_reset()
+{
+	m_pc = 1; // Yes, this is odd, but otherwise there are sync issues between writing and clearing PA in scv kungfurd when playing back adpcm.
+	m_sp = 0;
+	m_a = 0;
+	m_a_save = 0;
+	m_h = 0;
+	m_n = 0;
+	m_x = 0;
+	m_y = 0;
+	m_pa = 0;
+	m_pb = 0;
+	m_md0 = 0;
+	m_md1 = 0;
+	m_skip = false;
+	m_skip_save = false;
+	m_ts = false;
+	m_ns = false;
+	m_ss = false;
+	m_dac = 0;
+	m_dac_sign = true;
+	m_tn_int = false;
+	m_counter = 0;
+	m_timer = 0;
+	set_noise_counter_bit();
+}
+
+
+std::unique_ptr<util::disasm_interface> upd177x_cpu_device::create_disassembler()
+{
+	return std::make_unique<upd177x_disassembler>();
+}
+
+
+void upd177x_cpu_device::op0xxx(u16 opcode)
+{
+	switch (opcode)
+	{
+	case 0x0000:  // NOP
+		break;
+	case 0x0002:  // OUT PA
+		m_pa = m_a;
+		m_pa_out_cb(m_pa);
```

Comment:
There isn't much use for a callback here. There's no external indication of the 'OUT PA' at instruction execution time; the port pins don't immediately change. In order for the outside world to observe this instruction's effect, an external read (from the main CPU) must happen.

Reply:
You are right. I have removed the callback.